### PR TITLE
[AIRFLOW-297] support exponential backoff option for retry delay

### PR DESCRIPTION
**Pull request by jgao54**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
-[AIRFLOW-297](https://issues.apache.org/jira/browse/AIRFLOW-297)

Testing Done:
- Generated DAGs and ran tests locally
- Testing for this may extend total unit tests time a bit (given delay is exponential). If there are better ways for testing this, please let me know and I can add it as needed.

Reminders for contributors (REQUIRED!):
- Your PR's title must reference an issue on 
  [Airflow's JIRA](https://issues.apache.org/jira/browse/AIRFLOW/). 
  For example, a PR called "[AIRFLOW-1] My Amazing PR" would close JIRA 
  issue #1. Please open a new issue if required!
- Please squash your commits when possible and follow the [How to write a good git commit message](http://chris.beams.io/posts/git-commit/). 
  Summarized as follows:
  1. Separate subject from body with a blank line
  2. Limit the subject line to 50 characters
  3. Do not end the subject line with a period
  4. Use the imperative mood in the subject line (add, not adding)
  5. Wrap the body at 72 characters
  6. Use the body to explain what and why vs. how

### CODE REVIEW

**Comment by r39132** [comment]
Exponential back-off algos are best implemented as bounded, exponential back-off algos. In this case, the bound is provided by the `retries` configuration parameter on the Task (or default_args dict passed to the DAG and then to the tasks). For reasonable values of retries, the delay can grow pretty quickly to appear interminably high, so it might be a good idea to offer another optional parameter for max_retry_delay.

File: airflow/models.py [hunk 1]
```diff
@@ -1238,7 +1242,11 @@ def run(
             # todo: move this to the scheduler
                 self.state == State.UP_FOR_RETRY and
                 not self.ready_for_retry()):
-            next_run = (self.end_date + task.retry_delay).isoformat()
+            delay = task.retry_delay
```
Comment (criccomini) [inline airflow/models.py]: This code is copy/paste of the above. A helper function seems like a better approach.

**Comment by criccomini** [comment]
> it might be a good idea to offer another optional parameter for max_retry_delay

+1 This seems useful, and fairly straight forward to add (min(max_retry_delay, delay)).

Overall, PR looks good. Had a minor nit.

**Comment by criccomini** [comment]
LGTM

**Comment by criccomini** [comment]
Going to run this in our prod environment

**Pull request closed without merging**